cmd/encryption-v1.go

```diff
@@ -161,31 +181,44 @@ func EncryptRequest(content io.Reader, r *http.Request, metadata map[string]stri
 	delete(metadata, SSECustomerKey) // make sure we do not save the key by accident
 
 	// security notice:
-	// Reusing a tuple (nonce, client provided key) will produce the same encryption key
-	// twice and breaks the tamper-proof property. However objects are still confidential.
-	// Therefore the nonce must be unique but need not to be undistinguishable from true
-	// randomness.
-	nonce := make([]byte, 32) // generate random nonce to derive encryption key
+	//  - If the first 32 bytes of the random value are ever repeated under the same client-provided
+	//    key the encrypted object will not be tamper-proof. [ P (coll) ~= 1 / 2^(256 / 2)]
+	//  - If the last 32 bytes of the random value are ever repeated under the same client-provided
+	//    key an adversary may be able to extract the object encryption key. This depends on the
+	//    authenticated en/decryption scheme. The DARE format will generate an 8 byte nonce which must
+	//    be repeated in addition to reveal the object encryption key.
+	//    [ P[coll] ~= 1 / 2^((256 + 64) / 2) ]
+	nonce := make([]byte, 64) // generate random values for key derivation
 	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
 		return nil, err
 	}
-	iv := sha256.Sum256(nonce) // hash output to not reveal any stat. weaknesses of the PRNG
+	sha := sha256.New() // derive object encryption key
+	sha.Write(key)
+	sha.Write(nonce[:32])
+	objectEncryptionKey := sha.Sum(nil)
 
-	keyHash := sha256.Sum256(key) // derive MAC of the client-provided key
-	mac := hmac.New(sha256.New, keyHash[:])
-	mac.Write(iv[:])
-	keyMAC := mac.Sum(nil)
+	iv := sha256.Sum256(nonce[32:]) // derive key encryption key
+	sha = sha256.New()
+	sha.Write(key)
+	sha.Write(iv[:])
+	keyEncryptionKey := sha.Sum(nil)
 
-	mac = hmac.New(sha256.New, key) // derive encryption key
-	mac.Write(keyMAC)
-	reader, err := sio.EncryptReader(content, sio.Config{Key: mac.Sum(nil)})
+	sealedKey := bytes.NewBuffer(make([]byte, 0, 64)) // 16 byte header + 32 byte payload + 16 byte tag
+	n, err := sio.Encrypt(sealedKey, bytes.NewReader(objectEncryptionKey), sio.Config{
+		Key: keyEncryptionKey,
+	})
+	if n != 64 || err != nil {
+		return nil, errors.New("failed to seal object encryption key") // if this happens there's a bug in the code (may panic ?)
```

Comment:
or memory shortage?

Reply:
You mean the Go runtime runs OoM? That would cause a panic...

Reply:
Yes, I guess so. So let's leave it like this.